hera_cal/reflections.py

```diff
@@ -267,25 +239,445 @@ def write_reflections(self, output_calfits, input_calfits=None, overwrite=False,
+            self.uflags = DataContainer({})
+
+        # get selection function
+        if side == 'pos':
+            select = np.where((self.delays >= dly_range[0]) & (self.delays <= dly_range[1]))[0]
+        elif side == 'neg':
+            select = np.where((self.delays >= -dly_range[1]) & (self.delays <= -dly_range[0]))[0]
+        elif side == 'both':
+            select = np.where((np.abs(self.delays) > dly_range[0]) & (np.abs(self.delays) < dly_range[1]))[0]
+
+        # iterate over keys
+        for k in keys:
+            if k in self.svals and not overwrite:
+                echo("{} exists in svals and overwrite == False, skipping...".format(k), verbose=verbose)
+                continue
+            if k not in dfft:
+                echo("{} not found in dfft, skipping...".format(k), verbose=verbose)
+                continue
+
+            # perform svd to get principal components
+            d = np.zeros_like(dfft[k])
+            d[:, select] = dfft[k][:, select]
+            u, s, v = _svd_waterfall(d)
+            Nmodes = len(s)
+
+            # append to containers
+            self.umodes[k] = u[:, :Nmodes]
+            self.vmodes[k] = v[:Nmodes, :]
+            self.svals[k] = s
+            self.uflags[k] = np.min(flags[k], axis=1, keepdims=True)
+
+        # get principal components
+        self.form_PCs(keys, overwrite=overwrite)
+
+        # append relevant metadata
+        if hasattr(dfft, 'times'):
+            self.umodes.times = dfft.times
+            self.vmodes.times = dfft.times
+            self.svals.times = dfft.times
+            self.uflags.times = dfft.times
+
+    def form_PCs(self, keys=None, u=None, v=None, overwrite=False, verbose=True):
+        """
+        Build principal components from SVD u-modes and v-modes.
+
+        Take u and v-modes and form outer product to get principal components
+        and insert into self.pcomps
+
+        Args:
+            keys : list of tuples
+                List of baseline-pol DataContainer tuples to operate on.
+            u : DataContainer 
+                Holds u-modes to use in forming PCs. Default is self.umodes
+            v : DataContainer 
+                Holds v-modes to use in forming PCs. Default is self.vmodes
+            overwrite : bool
+                If True, overwrite output data if it exists.
+            verbose : bool
+                If True, report feedback to stdout.
+        """
+        if keys is None:
+            keys = self.svals.keys()
+        if u is None:
+            u = self.umodes
+        if v is None:
+            v = self.vmodes
+
+        if not hasattr(self, 'pcomps'):
+            self.pcomps = DataContainer({})
+
+        for k in keys:
+            if k in self.pcomps and not overwrite:
+                echo("{} exists in pcomps and overwrite == False, skipping...".format(k), verbose=verbose)
+                continue
+
+            self.pcomps[k] = _form_PCs(u[k], v[k])
```

Comment:
This seems like unnecessary memory bloat. You shouldn't ever have to store all the outer products. You should form them on the fly as necessary and sum them with the appropriate singular value weights.